File: leetcode/q525/n.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        presum = 0
        hm = {}
        hm[0] = -1
        ans = 0
        for i in range(len(nums)):
            if nums[i] == 0:
                presum -= 1
            else:
                presum += 1
            if presum in hm:
                ans = max(ans, i - hm[presum])
            else:
                hm[presum] = i
        return ans

File: nowcoder/nc41/n.cpp
#include<iostream>
#include<vector>
#include<cstdio>
#include<cstring>
#include<algorithm>
#include<unordered_map>

using namespace std;

class Solution {
public:
    int maxLength(vector<int>& arr) {
        unordered_map<int, int> hm;
        int i = 0, j  = 0;
        int ans = 0;
        for( ; j < arr.size(); j++) {
            if(hm.count(arr[j])) {
                i = max(i, hm[arr[j]] + 1);
            }
            hm[arr[j]] = j;
            ans = max(ans, j - i + 1);
        }
        return ans;
    }
};
